app/scripts/background.js

```diff
@@ -981,45 +981,88 @@ export function setupController(
           connectionStream: portStreamForCookieHandlerPage,
         });
       }
-      connectExternalExtension(remotePort);
+
+      const portStream =
+        overrides?.getPortStream?.(remotePort) || new PortStream(remotePort);
+
+      connectEip1193(portStream, remotePort.sender);
+
+      if (process.env.MULTICHAIN_API && isFirefox) {
+        const mux = setupMultiplex(portStream);
+        mux.ignoreStream(METAMASK_EIP_1193_PROVIDER);
+
+        connectCaipMultichain(
+          mux.createStream(
+            METAMASK_CAIP_MULTICHAIN_PROVIDER,
+            remotePort.sender,
+          ),
+        );
+      }
     }
   };
 
-  // communication with page or other extension
-  connectExternalExtension = (remotePort) => {
+  /**
+   * Connects a externally_connecatable Port to the MetaMask controller.
+   * This method identifies dapp clients and connects them differently from extension clients.
+   *
+   * @param {Port} remotePort - The port provided by a new context.
+   */
+  connectExternallyConnectable = (remotePort) => {
     const portStream =
       overrides?.getPortStream?.(remotePort) || new PortStream(remotePort);
+
+    const isDappConnecting = remotePort.sender.tab?.id;
+    if (isDappConnecting && process.env.MULTICHAIN_API) {
+      if (metamaskBlockedPorts.includes(remotePort.name)) {
+        return;
+      }
+
+      // this is triggered when a new tab is opened, or origin(url) is changed
+      if (remotePort.sender && remotePort.sender.tab && remotePort.sender.url) {
+        trackDappView(remotePort);
+      }
+
+      connectCaipMultichain(portStream, remotePort.sender);
+    } else {
+      connectEip1193(portStream, remotePort.sender);
+    }
+  };
+
+  /**
+   * Connects a Duplexstream to the MetaMask controller EIP-1193 API (via a multiplexed duplex stream).
+   *
+   * @param {DuplexStream} connectionStream - The duplex stream.
+   * @param sender - The remote port sender.
+   */
+  connectEip1193 = (connectionStream, sender) => {
     controller.setupUntrustedCommunicationEip1193({
-      connectionStream: portStream,
-      sender: remotePort.sender,
+      connectionStream,
+      sender,
     });
   };
 
-  connectExternalCaip = async (remotePort) => {
+  /**
+   * Connects a DuplexStream to the MetaMask controller Caip Multichain API.
+   *
+   * @param {DuplexStream} connectionStream - The duplex stream.
+   * @param sender - The remote port sender.
```

Comment:
Same thoughts as above.

```suggestion
   * @param {chrome.runtime.MessageSender} [sender] - The remote port sender.
```

Reply:
addressed in comment above